### PR TITLE
Add async wait methods that support custom timeout

### PR DESCRIPTION
**Pull request by jcavar**
This resolves https://github.com/ReactiveCocoa/ReactiveObjC/issues/95

### CODE REVIEW

**Comment by jcavar** [comment]
These failures doesn't seem to be related to changes so not sure if there is anything I can do about them?

File: ReactiveObjC/RACSignal.h [hunk 1]
```diff
@@ -461,6 +461,8 @@ typedef RACSignal * _Nullable (^RACSignalBindBlock)(ValueType _Nullable value, B
 ///
 /// Returns the first value received, or `defaultValue` if no value is received
 /// before the signal finishes or the method times out.
+- (nullable ValueType)asynchronousFirstOrDefault:(nullable ValueType)defaultValue success:(nullable BOOL *)success error:(NSError * _Nullable * _Nullable)error timeout:(NSTimeInterval)timeout;
+
 - (nullable ValueType)asynchronousFirstOrDefault:(nullable ValueType)defaultValue success:(nullable BOOL *)success error:(NSError * _Nullable * _Nullable)error;
```
Comment (erichoracek) [inline ReactiveObjC/RACSignal.h]: Can you add documentation comments to this methods? (the same as before is probably fine)

File: ReactiveObjC/RACSignal.h [hunk 2]
```diff
@@ -472,6 +474,8 @@ typedef RACSignal * _Nullable (^RACSignalBindBlock)(ValueType _Nullable value, B
 ///
 /// Returns whether the signal completed successfully before timing out. If NO,
 /// `error` will be set to any error that occurred.
+- (BOOL)asynchronouslyWaitUntilCompleted:(NSError * _Nullable * _Nullable)error timeout:(NSTimeInterval)timeout;
+
 - (BOOL)asynchronouslyWaitUntilCompleted:(NSError * _Nullable * _Nullable)error;
```
Comment (erichoracek) [inline ReactiveObjC/RACSignal.h]: Can you add documentation comments to this methods? (the same as before is probably fine)

File: ReactiveObjC/RACSignal.h [hunk 3]
```diff
@@ -461,6 +461,8 @@ typedef RACSignal * _Nullable (^RACSignalBindBlock)(ValueType _Nullable value, B
 ///
 /// Returns the first value received, or `defaultValue` if no value is received
 /// before the signal finishes or the method times out.
+- (nullable ValueType)asynchronousFirstOrDefault:(nullable ValueType)defaultValue success:(nullable BOOL *)success error:(NSError * _Nullable * _Nullable)error timeout:(NSTimeInterval)timeout;
+
 - (nullable ValueType)asynchronousFirstOrDefault:(nullable ValueType)defaultValue success:(nullable BOOL *)success error:(NSError * _Nullable * _Nullable)error;
 
 /// Spins the main run loop for a short while, waiting for the receiver to complete.
```
Comment (erichoracek) [inline ReactiveObjC/RACSignal.h]: Can you update this documentation comment to mention the provided timeout, e.g. "Spins the main run loop for a short while, waiting for the receiver to complete or the provided timeout to elapse."

File: ReactiveObjC/RACSignal.h [hunk 4]
```diff
@@ -461,6 +461,8 @@ typedef RACSignal * _Nullable (^RACSignalBindBlock)(ValueType _Nullable value, B
 ///
 /// Returns the first value received, or `defaultValue` if no value is received
 /// before the signal finishes or the method times out.
+- (nullable ValueType)asynchronousFirstOrDefault:(nullable ValueType)defaultValue success:(nullable BOOL *)success error:(NSError * _Nullable * _Nullable)error timeout:(NSTimeInterval)timeout;
```
Comment (erichoracek) [inline ReactiveObjC/RACSignal.h]: Can you update this method's documentation comment to mention the provided timeout, e.g. "Spins the main run loop for a short while, waiting for the receiver to send a `next` or the provided timeout to elapse."

**Comment by erichoracek** [comment]
Just a few comments about the documentation comments—thanks for the contribution!

**Comment by jcavar** [comment]
Cool, should be ready now I think

File: ReactiveObjC/RACSignal.h [hunk 5]
```diff
@@ -472,6 +489,18 @@ typedef RACSignal * _Nullable (^RACSignalBindBlock)(ValueType _Nullable value, B
 ///
 /// Returns whether the signal completed successfully before timing out. If NO,
 /// `error` will be set to any error that occurred.
+- (BOOL)asynchronouslyWaitUntilCompleted:(NSError * _Nullable * _Nullable)error timeout:(NSTimeInterval)timeout;
+
+/// Spins the main run loop for a short while, waiting for the receiver to complete
+/// or the provided timeout to elapse.
```
Comment (erichoracek) [inline ReactiveObjC/RACSignal.h]: It appears that this doc comment is for the wrong method—this method `asynchronouslyWaitUntilCompleted:` does not take "a provided timeout"—rather the one above this `asynchronouslyWaitUntilCompleted:timeout:` does.

Comment (jcavar) [inline ReactiveObjC/RACSignal.h]: Right ... should be fixed now.

**Pull request merged**